Implement SAF provider.

gilbsgilbs opened the pull request:
Late night project :night_with_stars: :disappointed:  …

**Changelog:**

- Implement a [SAF Document provider](https://developer.android.com/guide/topics/providers/create-document-provider). This allows third party apps to access files through RClone (e.g. password manager, Android's stock file manager).
  - :information_source: File download was implemented using `rclone cat` command.
  - :information_source: File upload was implemented using `rclone rcat` command.
- Rework build scripts so that project builds in Android Studio directly after `git clone`, without having to run specific gradle tasks under a specific set of environment variables.
- Fix other issues I had prior building (especially `google-services.json` being missing).
- Drop renames mapping in shared prefs; make rclone's config file the unique source of truth for remote names (rather than refreshing views everywhere across the app). Note that this will make users loose their existing renames. If you think this is a big deal, we could write some migration code.
- Other refactoring.

Hope you'll enjoy it.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/x0b/rcx?pullRequest=76) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/x0b/rcx?pullRequest=76) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/x0b/rcx?pullRequest=76) it.</sub>

x0b commented:
Oh man, I was going to release my attempt this week to the alpha testers 😀. So, here's the new plan:
1. I'm going to push my current workspace more often so that this doesn't happen any more.
2. I'm going ahead with releasing the alpha version
3. Then, we look into whether your version should replace it, or if we can merge the two ideas. Also, we'll see whether we can merge your other changes right away. 

However, I'm currently really distracted because of things that happened in my personal life, but I'm trying to keep things going.

gilbsgilbs commented:
Oh, that's my bad, I should have commented on the issue first to check if there were any progress. But I think there's a lot of improvements we can pick from this PR anyways, especially the rework on how rclone is built, the renames rework, and maybe pick some ideas from my SAF implementation.

x0b commented:
Yes, from taking a quick look:
- 5e4781bf918abf00f240d5b90dd22c12c0b08b3a: Looks good, although we cannot upgrade to API 30 until January, according to Google documentation
- 7904a6e2be5d4372d6be84a9b1bd9cbfa4396cd7 / e54ed51d6cf8b46266d6b1668a987dbc922f7260: Good (some changes required)
-  SAF: I'll have to look at this later

gilbsgilbs commented:
> 5e4781b: Looks good, although we cannot upgrade to API 30 until January, according to Google documentation

Yup sorry, fixed. Android Studio wasn't very clear about it.

I also added some migration code so that users will keep their renamed remotes.

gilbsgilbs commented:
Hey, I'm just passing by on this PR to update RClone and build tools. I've been using my fork for a while now and didn't notice any issue.

No urge to review though if you still have more important things to do :wink: . If you need help for something, do not hesitate. I'd love to help you get this app on F-Droid.

Take care.

x0b commented:
Sorry, I'm currently still very caught up with personal stuff. I'll try to get the 1.12 preview out this week and then look into your PR.
